### ISSUE TITLE
[LoongArch] dead loop after nightly-2024-02-12

### ISSUE THREAD

**Issue by heiher** [rust-lang/rust#121289]
I tried this code:

```rust
rustc -vV
cargo -vV
```

I expected to see this happen: outputs version info

Instead, this happened: dead loop in `main` entry function.

### Meta

`rustc --version --verbose`:
```
a84bb95a1 2024-02-13
```

```
b381d3ab2 2024-02-12 good
a84bb95a1 2024-02-13 bad
```

**Comment by nikic** [comment]
That would be due to the LLVM 18 upgrade.

**Comment by heiher** [comment]
There is no version of crosstool-ng that supports `R_LARCH_CALL36`, I'll revert the [`medium` code model](https://github.com/rust-lang/rust/pull/120661) until the issue is resolved.

**Comment by xry111** [comment]
Oooops.  Is this communicated with crosstool-ng maintainers?

**Comment by xry111** [comment]
But maybe crosstool-ng maintainers can't act as well due to the incompatibility between GCC 13.2 and Binutils 2.42 (loongson-community/discussions#41) :(.  I'm not sure how to fix such an incompatibility...

**Comment by xen0n** [comment]
https://github.com/rust-lang/rust/pull/129048 is merged recently that bumped the LoongArch binutils to 2.42, so can we try #120661 again?

**Comment by heiher** [comment]
> #129048 is merged recently that bumped the LoongArch binutils to 2.42, so can we try #120661 again?

Yes, with #129048 merged, we can retry #120661.

**Comment by xen0n** [comment]
Today's nightly is built off of #130312, which contains #130266 and does work:

```
$ rustc -Vv
rustc 1.83.0-nightly (0609062a9 2024-09-13)
binary: rustc
commit-hash: 0609062a91c8f445c3e9a0de57e402f9b1b8b0a7
commit-date: 2024-09-13
host: loongarch64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0
```

And the `medium` code model is indeed in use:

```
$ readelf -r ~/.rustup/toolchains/nightly-loongarch64-unknown-linux-gnu/lib/rustlib/loongarch64-unknown-linux-gnu/lib/libcore-c292f7f95e655c4c.rlib | grep 'R_LARCH_CALL36' | wc -l
769
```

So this issue is finally resolved. Thanks for everyone involved!
